### PR TITLE
build: replace prettier with biome

### PR DESCRIPTION
**Pull request by anonrig**
# Summary

This PR replaces Prettier and several Eslint rules with Biome for reduced CI times for faster linting & formatting. **Linting and formatting now takes 44 seconds, compared to `master` branch with 95 seconds.** 

Here's a list of caveats with the proposed changes:

- Biome has 2 unique bugs which led to adding `biome-ignore format` in several files. The issues are tracked on Biome repository.
  - https://github.com/biomejs/biome/issues/1106
  - https://github.com/biomejs/biome/issues/1112
- `eslint-config-sentry` requires and installs `prettier`. Therefore, removing from `sentry` repository package.json causes an error. We eventually need to make `prettier` optional in the upstream repository. 

# Why Biome?

I recommend [reading the Philosophy behind Biome](https://biomejs.dev/internals/philosophy/).

- Biome has a [built-in linter](https://biomejs.dev/linter/) (similar to Eslint) and [has tons of lint rules.](https://biomejs.dev/linter/rules/)
- Biome [has a formatter](https://biomejs.dev/formatter/) which is ~95% similar to Prettier. There are some differences between Prettier that's [documented here](https://biomejs.dev/formatter/differences-with-prettier/)
- Biome has a [built-in import sorting/ordering](https://biomejs.dev/analyzer/)
- Biome has a [built-in LSP, VScode and IntelliJ extension.](https://biomejs.dev/reference/vscode/). It [has third-part plugins](https://biomejs.dev/guides/integrate-in-editor/#third-party-plugins) for nevoid, helix, doc.nvim and sublime text.
- Biome [has a Github actions](https://biomejs.dev/recipes/continuous-integration/) that doesn't depend on Node.js to be run.
- Biome supports JS, TS, JSX, JSON, JSONC and CSS. And has Markdown and CSS [in their roadmap.](https://biomejs.dev/internals/language-support/)
- Biome supports extensible/shareable configuration files. Similar to `eslint-config-sentry` [we should share Biome config.](https://biomejs.dev/guides/big-projects/)
- Biome supports [pre-commit hooks.](https://github.com/biomejs/pre-commit)

# Impact

> :warning: **The following results are taken on my work M2 machine.** The results will be a lot different on shared VM such as GitHub Actions. Please use this as a baseline and execute the following scripts on your own machine.

### Duration

It takes 95 seconds to run `yarn lint` which includes prettier with eslint. In order

```
yarn lint  95.66s user 4.21s system 159% cpu 1:02.74 total
```

### Time consuming Eslint rules

The following shows the top 10 most time spent rules in Eslint. In order to reproduce the following on your local, please run `TIMING=10 yarn lint`

```
Rule                                                          | Time (ms) | Relative
:-------------------------------------------------------------|----------:|--------:
prettier/prettier                                             | 21383.555 |    45.6%
import/default                                                |  6131.662 |    13.1%
import/no-unresolved                                          |  3002.880 |     6.4%
@typescript-eslint/naming-convention                          |  2618.324 |     5.6%
react/no-direct-mutation-state                                |  1669.827 |     3.6%
no-lookahead-lookbehind-regexp/no-lookahead-lookbehind-regexp |  1287.728 |     2.7%
import/no-nodejs-modules                                      |  1058.123 |     2.3%
@typescript-eslint/no-redeclare                               |   889.496 |     1.9%
react/sort-comp                                               |   599.328 |     1.3%
react/no-typos                                                |   578.446 |     1.2%
```

### Duration

It takes 44 seconds to run `yarn lint` command, and only 531ms is spent on Biome. Ideally moving more linting rules to Biome will reduce the linting speed.

```
yarn lint  44.17s user 2.54s system 151% cpu 30.738 total
```

### Remaining Eslint rules

The following shows the top 10 most time spent rules in Eslint. In order to reproduce the following on your local, please run `TIMING=10 yarn lint:js`

```
Rule                                                          | Time (ms) | Relative
:-------------------------------------------------------------|----------:|--------:
import/no-duplicates                                          |  3262.210 |    23.3%
@typescript-eslint/naming-convention                          |  2395.588 |    17.1%
react/no-direct-mutation-state                                |  1504.342 |    10.8%
no-lookahead-lookbehind-regexp/no-lookahead-lookbehind-regexp |  1127.794 |     8.1%
react/sort-comp                                               |   608.894 |     4.4%
react/function-component-definition                           |   491.970 |     3.5%
simple-import-sort/imports                                    |   455.786 |     3.3%
react/jsx-fragments                                           |   297.780 |     2.1%
jest/no-large-snapshots                                       |   252.458 |     1.8%
react/no-deprecated                                           |   187.937 |     1.3%
```

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -38,6 +51,10 @@ module.exports = {
   },
 
   overrides: [
+    {
+      files: ['test/js/**/*.{ts,js}'],
```
[the remark this answers is not in the document]
Comment (scttcper) [reply .eslintrc.js:55]: i guess the sentry-app config would probably already do that nevermind

[the remark this answers is not in the document]
Comment (anonrig) [reply .eslintrc.js:55]: Yes, this change is only required for `tests/js` subfolder.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -112,32 +112,6 @@ repos:
...
         args: [--pytest-test-first]
-  - repo: https://github.com/pre-commit/mirrors-prettier
-    rev: 'v3.0.3' # Use the sha or tag you want to point at
-    hooks:
-      - id: prettier
-        name: prettier (yaml, json, markdown)
-        types_or:
-          - json
-          - yaml
-          - markdown
```
Comment (asottile-sentry) [inline .pre-commit-config.yaml:123]: does biome format json / yaml / markdown? I was under the impression it was mainly targetted at javascript

Comment (anonrig) [reply .pre-commit-config.yaml:123]: Here's a list of supported languages: https://biomejs.dev/internals/language-support/

**Review by asottile-sentry — APPROVED** [review]

**Comment by mitsuhiko** [comment]
Is there a reason we did not merge this?

**Comment by anonrig** [comment]
> Is there a reason we did not merge this? (Or the other one)

I couldn't find anybody to land this and later help me track the health of the release, in case there is a bug.

**Pull request closed without merging**